Add MAUI example app to csharp samples

weslleymurdock opened the pull request:
This covers #4077 
I've created the sample application for both iOS & Android platforms, and some modifications on makefile to generate c# PINVOKE classes to the right project as well

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pjsip/pjproject?pullRequest=4078) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pjsip/pjproject?pullRequest=4078) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pjsip/pjproject?pullRequest=4078) it.</sub>

weslleymurdock commented:
Hello
Which version of dotnet you`re using 
Please post the dotnet setup details

weslleymurdock commented:
Hi again
I'll try to create a global.json to configure the .ned sdk version as the same i've used

weslleymurdock commented:
Hello @trengginas  I've reproduced this when i'm used the global on a mac. And i've done the android sample on windows, which uses the ResolveNativeReference MSBuild Task for resolve the android libs. Try use windows while i'm working on a new android sample

trengginas commented:
I'm using a Mac since it's not possible to build the iOS sample app on Windows.

weslleymurdock commented:
To "build on windows" you can use the ssh connection of visual studio to debug and run the iOS app on emulator inside Windows Environment.
But if is not possible use a windows pc, try to build one target only (e.g Android only or iOS only).

weslleymurdock commented:
@sauwming, for sure, i'll work on that today

sauwming commented:
Thanks, it would be better if the sample apps can work on real devices, instead of simulators. If I tried following `Readme.md`, I would encounter the following error:
`building for iOS Simulator, but linking in object file built for iOS`

weslleymurdock commented:
The steps for real devices are similar, but I'll improve the readme this week

weslleymurdock commented:
@sauwming I figure it out. The project has a conditional compilation in order to split platforme codes and resources
And the inclusion of arm64 library is included in x86_64, causing the bad linking
I'll updathe the pr with a including  conditional check on arm64 lib

weslleymurdock commented:
Hi fowks
Im working on desktop maui samples, so i maked this as draft
It works on windows, but it still need the native platform controls implementation for video win

sauwming commented:
I still haven't managed to make it work on a real device.

On iOS, it will fail due to an extra opening parenthesis here:
`<ItemGroup Condition="(($(RuntimeIdentifier.StartsWith('iossimulator-arm64')))">`

But even after I fixed it, it still failed with "undefined references" errors.

weslleymurdock commented:
that's is wird... the arch on condition is for simulators running on M1/M2 macs
the undefined references is about linking
could you please post me a build with diag in order to see whats going on with linking?

sauwming commented:
I finally managed to make it work. I have to specify the `RuntimeIdentifier`, otherwise it will link with simulator.

`dotnet publish -c Debug -f net9.0-ios -p:RuntimeIdentifier=ios-arm64`

The result is an `.ipa` (iOS app package) in `pjsip-apps/src/swig/csharp/pjsua2maui/pjsua2maui/bin/Debug/net9.0-ios/ios-arm64/publish/pjsua2maui.ipa`, which I can then deploy to the device using Xcode. Not sure if there is an easier way to do this.

weslleymurdock commented:
When i wrote the sample, i used the .NET 8.0 and it runs fine on android. However, i've upgraded to .NET 9.0 and didn't tested it. 

I've debuging the android app on VS using windows, and it shows the crash when the app set's the camera manager by invoking the Java code injected on the app., but it seems that the java code does not being injected to app when putting the files on project app and flagging it as ```<AndroidJavaSource>SomeFile.java</AndroidJavaSource>``` .

I'm currently writing a softphone for a company, and faced this issue. My workaround was create a Class Library Project, and produce a library. This way the file is injected to final binary, and the app can handle the PjCamera2 methods loading for set the CameraManager.
 
I don't know if there is a issue with .NET 9.0, but i can provide a new solution, where the app consumes the wrapper classes and native libraries from a .net maui library project, or even a local nuget package.

weslleymurdock commented:
Hi folks
Sorry by the late answer
I've update the version of dotnet sdk, and build the library and tested the project. It seems that the bug on dotnet that was causing the crash on the app initialization was fixed
I've fixed one line of android code where the swig generator dont generate one method, and updated the pr

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
```diff
@@ -0,0 +1,126 @@
+    - Visual Studio Code & .NET MAUI - [Install steps](https://learn.microsoft.com/en-us/dotnet/maui/get-started/installation?view=net-maui-11.0&tabs=visual-studio-code)
+    - Java 17 
+    - Android NDK r21d
+    - pjsip v2.14.1
```
Change this to:
- Android NDKr21d or later
- pjsip v2.14.1 or later

Here I'm using NDK r28 and 2.15.1, and it seems ok

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
````diff
@@ -0,0 +1,126 @@
+
+```bash
+
+TARGET_ABI=x86_64  ./configure-android
````
Need instruction for device too, i.e.:
For device:
`./configure-android`
For simulator:
`TARGET_ABI=x86_64 ./configure-android`

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
```diff
@@ -0,0 +1,126 @@
+    sdkmanager --install "ndk;28.0.12674087" --channel=3 --sdk_root=$ANDROID_SDK_ROOT
+    sdkmanager --install "cmake;3.31.0" --sdk_root=$ANDROID_SDK_ROOT
+    sdkmanager --install "build-tools;35.0.0" "platform-tools" "platforms;android-35" --sdk_root=$ANDROID_SDK_ROOT
+    echo yes | sdkmanager --licenses --sdk_root=$ANDROID_SDK_ROOT
```
Add note:
If you are sharing the Android SDK location with Android Studio, the recommended way to install/update the SDK and accept licenses is by using Android Studio.

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
````diff
@@ -0,0 +1,126 @@
+
+```
+
+then start the android emulator, go to ``` pjsip-apps/src/swig/csharp/pjsua2maui/pjsua2maui ``` and run:
````
This also needs instruction for device.

Here's what I did:
For Android device:
* Open Android Studio and connect an Android device, such as by using "Pair Devices Using Wifi"
* (Optional) Start Logcat in VS Code, for obtaining Android log.
   - Install Logcat: View-Extensions. Install Logcat extension.
   - Start Logcat: View-Command Pallette.

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
```diff
@@ -0,0 +1,126 @@
+- SETUP:
+    - Visual Studio Code & .NET MAUI - [Install steps](https://learn.microsoft.com/en-us/dotnet/maui/get-started/installation?view=net-maui-11.0&tabs=visual-studio-code)
+    - Xcode 16.1 
+    - pjsip v2.14.1
```
same as above. add "or later"

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
```diff
@@ -0,0 +1,126 @@
+
+#define PJ_CONFIG_IPHONE 1
+
+// disable background VoIP socket, use PushKit
```
These two defines below should not be needed.

The setting to disable background voip is: `PJ_ACTIVESOCK_TCP_IPHONE_OS_BG`, which is by default 0.
No need to define min iOS version as well.

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
````diff
@@ -0,0 +1,126 @@
+
+```bash
+
+export DEVPATH="`xcrun -sdk iphonesimulator --show-sdk-platform-path`/Developer"
````
Move this below

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
````diff
@@ -0,0 +1,126 @@
+
+```bash
+
+MIN_IOS="-miphoneos-version-min=12.2" ARCH="-arch x86_64" CFLAGS="-O2 -m32 -mios-simulator-version-min=12.2 -fembed-bitcode" LDFLAGS="-O2 -m32 -mios-simulator-version-min=12.2 -fembed-bitcode" ./configure-iphone
````
Need instruction for device as well.

For device:
`./configure-iphone`
For simulator:
`export DEVPATH="`xcrun -sdk iphonesimulator --show-sdk-platform-path`/Developer"`
...

sauwming commented on pjsip-apps/src/swig/csharp/pjsua2maui/README.md:
```diff
@@ -0,0 +1,126 @@
+
+dotnet workload restore
+dotnet restore
+dotnet build -t:Run -c Debug -f net9.0-ios -p:_DeviceName=:v2:udid=UDID_OF_YOUR_EMULATOR
```
Need instruction for device as well.
With a real device, the code signing is more complicated using VSCode, so here's what I did:

- dotnet publish -c Debug -f net9.0-ios -p:RuntimeIdentifier=ios-arm64
- Open the resulting `.ipa` file in Xcode: `pjsip-apps/src/swig/csharp/pjsua2maui/pjsua2maui/bin/Debug/net9.0-ios/ios-arm64/publish/pjsua2maui.ipa`
Drag it to your connected device under "Devices and Simulators."
Xcode will handle the signing and deployment.

sauwming reviewed (commented):
I tested it on Android and it no longer crashes during startup.

Need more detailed instruction though on how to run the project on real mobile devices, so you need to modify `README.md`.
